test/grouping.jl

```diff
@@ -86,89 +85,237 @@ module TestGrouping
-    gd = groupby(df, cols, sort=true)
-    ga = map(f, gd)
-    @test sbdf == combine(ga)
-
-    g(df) = DataFrame(cmax1 = [c + 1 for c in df[:cmax]])
-    h(df) = g(f(df))
-
-    @test combine(map(h, gd)) == combine(map(g, ga))
-
-    # testing pool overflow
-    df2 = DataFrame(v1 = categorical(collect(1:1000)), v2 = categorical(fill(1, 1000)))
-    @test groupby(df2, [:v1, :v2]).starts == collect(1:1000)
-    @test groupby(df2, [:v2, :v1]).starts == collect(1:1000)
-
-    # grouping empty table
-    @test groupby(DataFrame(A=Int[]), :A).starts == Int[]
-    # grouping single row
-    @test groupby(DataFrame(A=Int[1]), :A).starts == Int[1]
-
-    # testing pool overflow
-    df2 = DataFrame(v1 = categorical(collect(1:1000)), v2 = categorical(fill(1, 1000)))
-    @test groupby(df2, [:v1, :v2]).starts == collect(1:1000)
-    @test groupby(df2, [:v2, :v1]).starts == collect(1:1000)
-
-    # grouping empty table
-    @test groupby(DataFrame(A=Int[]), :A).starts == Int[]
-    # grouping single row
-    @test groupby(DataFrame(A=Int[1]), :A).starts == Int[1]
-
-    # issue #960
-    x = CategoricalArray(collect(1:20))
-    df = DataFrame(v1=x, v2=x)
-    groupby(df, [:v1, :v2])
-
-    df2 = by(e->1, DataFrame(x=Int64[]), :x)
-    @test size(df2) == (0, 1)
-    @test sum(df2[:x]) == 0
-
-    # Check that reordering levels does not confuse groupby
-    df = DataFrame(Key1 = CategoricalArray(["A", "A", "B", "B"]),
-                   Key2 = CategoricalArray(["A", "B", "A", "B"]),
-                   Value = 1:4)
-    gd = groupby(df, :Key1)
-    @test gd[1] == DataFrame(Key1=["A", "A"], Key2=["A", "B"], Value=1:2)
-    @test gd[2] == DataFrame(Key1=["B", "B"], Key2=["A", "B"], Value=3:4)
-    gd = groupby(df, [:Key1, :Key2])
-    @test gd[1] == DataFrame(Key1="A", Key2="A", Value=1)
-    @test gd[2] == DataFrame(Key1="A", Key2="B", Value=2)
-    @test gd[3] == DataFrame(Key1="B", Key2="A", Value=3)
-    @test gd[4] == DataFrame(Key1="B", Key2="B", Value=4)
-    # Reorder levels, add unused level
-    levels!(df[:Key1], ["Z", "B", "A"])
-    levels!(df[:Key2], ["Z", "B", "A"])
-    gd = groupby(df, :Key1)
-    @test gd[1] == DataFrame(Key1=["A", "A"], Key2=["A", "B"], Value=1:2)
-    @test gd[2] == DataFrame(Key1=["B", "B"], Key2=["A", "B"], Value=3:4)
-    gd = groupby(df, [:Key1, :Key2])
-    @test gd[1] == DataFrame(Key1="A", Key2="A", Value=1)
-    @test gd[2] == DataFrame(Key1="A", Key2="B", Value=2)
-    @test gd[3] == DataFrame(Key1="B", Key2="A", Value=3)
-    @test gd[4] == DataFrame(Key1="B", Key2="B", Value=4)
+    @testset "by and groupby" begin
+        Random.seed!(1)
+        df = DataFrame(a = repeat(Union{Int, Missing}[1, 3, 2, 4], outer=[2]),
+                       b = repeat(Union{Int, Missing}[2, 1], outer=[4]),
+                       c = Vector{Union{Float64, Missing}}(randn(8)))
+
+        cols = [:a, :b]
+        f(df) = DataFrame(cmax = maximum(df[:c]))
+        g(df) = (cmax = maximum(df[:c]),)
+        h(df) = maximum(df[:c])
```

Comment:
Do we test against: `Vector`, `Matrix`, `SubDataFrame` and `DataFrameRow`?
Also testing for more than one column might be good.
Do we test against a case when what `f` returns produces a duplicate column name?

Reply:
`SubDataFrame` is tested via `identity`. I've added tests for other cases. Duplicated names were exercised by `identity` too, but I've adapted the code to really check the names of duplicated columns.